File: src/proj4/projections/omerc.ts
import { ProjectionBase } from '.';
import { D2R, EPSLN, HALF_PI, QUART_PI, TWO_PI } from '../constants';
import { adjustLon, phi2z, tsfnz } from '../common';

import type { VectorPoint } from '../../geometry';
import type { ProjectionParams, ProjectionTransform } from '.';

const { abs, pow, sin, cos, sqrt, atan2, asin, log, atan, PI, tan, exp } = Math;

/**
 * # Oblique Mercator
 *
 * **Classification**: Conformal cylindrical
 *
 * **Available forms**: Forward and inverse, spherical and ellipsoidal
 *
 * **Defined area**: Global, but reasonably accurate only within 15 degrees of the oblique central line
 *
 * **Alias**: omerc
 *
 * **Domain**: 2D
 *
 * **Input type**: Geodetic coordinates
 *
 * **Output type**: Projected coordinates
 *
 * ## Projection String
 * ```
 * +proj=omerc +lat_1=45 +lat_2=55
 * ```
 *
 * ## Required Parameters
 * - `+lat_1=<value>`: Latitude of the first point on the central line.
 * - `+lat_2=<value>`: Latitude of the second point on the central line.
 *
 * ## Optional Parameters
 * - `+alpha=<value>`: Azimuth of the centerline clockwise from north at the center point of the line.
 * - `+gamma=<value>`: Azimuth of the centerline clockwise from north of the rectified bearing of the centerline.
 * - `+lonc=<value>`: Longitude of the projection center (overrides `+lon_0`).
 * - `+lat_0=<value>`: Latitude of the projection center.
 * - `+no_rot`: Disables rectification (historical reason).
 * - `+no_off`: Disables origin offset to the center of projection.
 * - `+k_0=<value>`: Scale factor at the central line.
 * - `+x_0=<value>`: False easting.
 * - `+y_0=<value>`: False northing.
 *
 * ## Usage Example
 * ```
 * echo 12 55 | proj +proj=omerc +alpha=90 +ellps=GRS80
 * echo 12 55 | proj +proj=omerc +alpha=0 +R=6400000
 * echo 12 55 | proj +proj=omerc +lon_1=0 +lat_1=-1 +lon_2=0 +lat_2=0 +R=6400000
 * echo 12 55 | proj +proj=tmerc +R=6400000
 * echo 12 55 | proj +proj=omerc +lon_1=-1 +lat_1=1 +lon_2=0 +lat_2=0 +ellps=GRS80
 * echo 10.536498003 56.229892362 | cs2cs +proj=longlat +ellps=GRS80 +to +proj=omerc +axis=wnu +lonc=9.46 +lat_0=56.13333333 +x_0=-266906.229 +y_0=189617.957 +k=0.9999537 +alpha=-0.76324 +gamma=0 +ellps=GRS80
 * ```
 *
 * ## Caveats
 * The two-point method with no rectification is probably only marginally useful.
 *
 * ![Oblique Mercator](https://github.com/Open-S2/s2-tools/blob/master/assets/proj4/projections/images/omerc.png?raw=true)
 */
export class HotineObliqueMercator extends ProjectionBase implements ProjectionTransform {
  name = 'HotineObliqueMercator';
  static names = [
    'HotineObliqueMercator',
    'Hotine_Oblique_Mercator',
    'Hotine Oblique Mercator',
    'Hotine_Oblique_Mercator_Azimuth_Natural_Origin',
    'Hotine_Oblique_Mercator_Two_Point_Natural_Origin',
    'Hotine_Oblique_Mercator_Azimuth_Center',
    'Oblique_Mercator',
    'omerc',
  ];
  // HotineObliqueMercator specific variables
  noOff: boolean;
  noRot: boolean;
  TOL = 1e-7;
  lam0: number;
  declare long2: number;
  declare longc: number;
  A = 0;
  B = 0;
  E = 0;
  singam: number;
  cosgam: number;
  sinrot: number;
  cosrot: number;
  rB: number;
  ArB: number;
  BrA: number;
  u0: number;
  vPoleN: number;
  vPoleS: number;

  /**
   * Preps an HotineObliqueMercator projection
   * @param params - projection specific parameters
   */
  constructor(params?: ProjectionParams) {
    super(params);
    const { TOL } = this;

    let con,
      cosph0,
      D,
      F,
      H,
      L,
      sinph0,
      p,
      J,
      gamma = 0,
      gamma0,
      lamc = 0,
      lam1 = 0,
      lam2 = 0,
      phi1 = 0,
      phi2 = 0,
      alphaC = 0;

    // only Type A uses the noOff or noUoff property
    // https://github.com/OSGeo/proj.4/issues/104
    this.noOff = isTypeA(params ?? {});
    this.noRot = params?.noRot ?? false;

    let alp = false;
    if ('alpha' in this) {
      alp = true;
    }

    let gam = false;
    if (this.rectifiedGridAngle !== undefined) {
      gam = true;
      gamma = this.rectifiedGridAngle * D2R;
    }

    if (alp) {
      alphaC = this.alpha ?? 0;
    }

    if (alp || gam) {
      lamc = this.longc;
    } else {
      lam1 = this.long1;
      phi1 = this.lat1;
      lam2 = this.long2;
      phi2 = this.lat2;

      if (
        abs(phi1 - phi2) <= TOL ||
        (con = abs(phi1)) <= TOL ||
        abs(con - HALF_PI) <= TOL ||
        abs(abs(this.lat0) - HALF_PI) <= TOL ||
        abs(abs(phi2) - HALF_PI) <= TOL
      ) {
        throw new Error();
      }
    }

    const one_es = 1.0 - this.es;
    const com = sqrt(one_es);

    if (abs(this.lat0) > EPSLN) {
      sinph0 = sin(this.lat0);
      cosph0 = cos(this.lat0);
      con = 1 - this.es * sinph0 * sinph0;
      this.B = cosph0 * cosph0;
      this.B = sqrt(1 + (this.es * this.B * this.B) / one_es);
      this.A = (this.B * this.k0 * com) / con;
      D = (this.B * com) / (cosph0 * sqrt(con));
      F = D * D - 1;

      if (F <= 0) {
        F = 0;
      } else {
        F = sqrt(F);
        if (this.lat0 < 0) {
          F = -F;
        }
      }

      this.E = F += D;
      this.E *= pow(tsfnz(this.e, this.lat0, sinph0), this.B);
    } else {
      this.B = 1 / com;
      this.A = this.k0;
      this.E = D = F = 1;
    }

    if (alp || gam) {
      if (alp) {
        gamma0 = asin(sin(alphaC) / D);
        if (!gam) {
          gamma = alphaC;
        }
      } else {
        gamma0 = gamma;
        alphaC = asin(D * sin(gamma0));
      }
      this.lam0 = lamc - asin(0.5 * (F - 1 / F) * tan(gamma0)) / this.B;
    } else {
      H = pow(tsfnz(this.e, phi1, sin(phi1)), this.B);
      L = pow(tsfnz(this.e, phi2, sin(phi2)), this.B);
      F = this.E / H;
      p = (L - H) / (L + H);
      J = this.E * this.E;
      J = (J - L * H) / (J + L * H);
      con = lam1 - lam2;

      if (con < -PI) {
        lam2 -= TWO_PI;
      } else if (con > PI) {
        lam2 += TWO_PI;
      }

      this.lam0 = adjustLon(
        0.5 * (lam1 + lam2) - atan((J * tan(0.5 * this.B * (lam1 - lam2))) / p) / this.B,
      );
      gamma0 = atan((2 * sin(this.B * adjustLon(lam1 - this.lam0))) / (F - 1 / F));
      gamma = alphaC = asin(D * sin(gamma0));
    }

    this.singam = sin(gamma0);
    this.cosgam = cos(gamma0);
    this.sinrot = sin(gamma);
    this.cosrot = cos(gamma);

    this.rB = 1 / this.B;
    this.ArB = this.A * this.rB;
    this.BrA = 1 / this.ArB;

    if (this.noOff) {
      this.u0 = 0;
    } else {
      this.u0 = abs(this.ArB * atan(sqrt(D * D - 1) / cos(alphaC)));

      if (this.lat0 < 0) {
        this.u0 = -this.u0;
      }
    }

    F = 0.5 * gamma0;
    this.vPoleN = this.ArB * log(tan(QUART_PI - F));
    this.vPoleS = this.ArB * log(tan(QUART_PI + F));
  }

  /**
   * HotineObliqueMercator forward equations--mapping lon-lat to x-y
   * @param p - lon-lat WGS84 point
   */
  forward(p: VectorPoint): void {
    let S, T, U, V, W, temp, u, v;
    p.x = p.x - this.lam0;
    if (abs(abs(p.y) - HALF_PI) > EPSLN) {
      W = this.E / pow(tsfnz(this.e, p.y, sin(p.y)), this.B);
      temp = 1 / W;
      S = 0.5 * (W - temp);
      T = 0.5 * (W + temp);
      V = sin(this.B * p.x);
      U = (S * this.singam - V * this.cosgam) / T;
      if (abs(abs(U) - 1.0) < EPSLN) {
        throw new Error();
      }
      v = 0.5 * this.ArB * log((1 - U) / (1 + U));
      temp = cos(this.B * p.x);
      if (abs(temp) < this.TOL) {
        u = this.A * p.x;
      } else {
        u = this.ArB * atan2(S * this.cosgam + V * this.singam, temp);
      }
    } else {
      v = p.y > 0 ? this.vPoleN : this.vPoleS;
      u = this.ArB * p.y;
    }
    if (this.noRot) {
      p.x = u;
      p.y = v;
    } else {
      u -= this.u0;
      p.x = v * this.cosrot + u * this.sinrot;
      p.y = u * this.cosrot - v * this.sinrot;
    }
    p.x = this.a * p.x + this.x0;
    p.y = this.a * p.y + this.y0;
  }

  /**
   * HotineObliqueMercator inverse equations--mapping x-y to lon-lat
   * @param p - HotineObliqueMercator point
   */
  inverse(p: VectorPoint): void {
    let u, v;
    p.x = (p.x - this.x0) * (1.0 / this.a);
    p.y = (p.y - this.y0) * (1.0 / this.a);
    if (this.noRot) {
      v = p.y;
      u = p.x;
    } else {
      v = p.x * this.cosrot - p.y * this.sinrot;
      u = p.y * this.cosrot + p.x * this.sinrot + this.u0;
    }
    const Qp = exp(-this.BrA * v);
    const Sp = 0.5 * (Qp - 1 / Qp);
    const Tp = 0.5 * (Qp + 1 / Qp);
    const Vp = sin(this.BrA * u);
    const Up = (Vp * this.cosgam + Sp * this.singam) / Tp;
    if (abs(abs(Up) - 1) < EPSLN) {
      p.x = 0;
      p.y = Up < 0 ? -HALF_PI : HALF_PI;
    } else {
      p.y = this.E / sqrt((1 + Up) / (1 - Up));
      p.y = phi2z(this.e, pow(p.y, 1 / this.B));
      if (p.y === Infinity) {
        throw new Error();
      }
      p.x = -this.rB * atan2(Sp * this.cosgam - Vp * this.singam, cos(this.BrA * u));
    }
    p.x += this.lam0;
  }
}

/**
 * @param P - projection parameters
 * @returns - true if projection is of type A
 */
function isTypeA(P: ProjectionParams): boolean {
  const typeAProjections = [
    'Hotine_Oblique_Mercator',
    'Hotine_Oblique_Mercator_Azimuth_Natural_Origin',
  ];

  return (
    'noUoff' in P ||
    'noOff' in P ||
    typeAProjections.indexOf(P.PROJECTION ?? P.name ?? 'XXXXXX') !== -1
  );
}

File: rust/proj/project/omerc.rs
use crate::proj::{
    aasin, adjlon, phi2, tsfn, CoordinateStep, Proj, ProjValue, ProjectCoordinates,
    TransformCoordinates, ANGLE_RECTIFIED_TO_SKEW_GRID, AZIMUTH_PROJECTION_CENTRE,
    HOTINE_OBLIQUE_MERCATOR_VARIANT_A, HOTINE_OBLIQUE_MERCATOR_VARIANT_B, LATITUDE_OF_FIRST_POINT,
    LATITUDE_OF_SECOND_POINT, LONGITUDE_OF_FIRST_POINT, LONGITUDE_OF_PROJECTION_CENTRE,
    LONGITUDE_OF_SECOND_POINT, NO_OFF, NO_ROTATION, NO_UOFF
};
use alloc::rc::Rc;
use core::{
    cell::RefCell,
    f64::consts::{FRAC_PI_2, FRAC_PI_4, PI, TAU},
};
use libm::{atan, atan2, cos, exp, fabs, log, pow, sin, sqrt, tan};

/// Oblique Mercator Variables
/// INTERNA
#[derive(Debug, Default, Clone, PartialEq)]
pub struct OmercData {
    a: f64,
    b: f64,
    e: f64,
    ab: f64,
    ar_b: f64,
    br_a: f64,
    r_b: f64,
    singam: f64,
    cosgam: f64,
    sinrot: f64,
    cosrot: f64,
    v_pole_n: f64,
    v_pole_s: f64,
    u_0: f64,
    no_rot: bool,
}

const TOL: f64 = 1e-7;
const EPS: f64 = 1e-10;

/// Hotine Oblique Mercator (variant A) Projection
/// EPSG Codes Used by Hotine Oblique Mercator (variant A): 8811, 8812, 8813, 8814, 8815, 8806, 8807
pub type HotineObliqueMercatorVariantAProjection = ObliqueMercatorProjection<HOTINE_OBLIQUE_MERCATOR_VARIANT_A>;
/// Hotine Oblique Mercator (variant B) Projection
/// EPSG Codes Used by Hotine Oblique Mercator (variant B): 8811, 8812, 8813, 8814, 8815, 8816, 8817
pub type HotineObliqueMercatorVariantBProjection = ObliqueMercatorProjection<HOTINE_OBLIQUE_MERCATOR_VARIANT_B>;

/// Oblique Mercator Projection
#[derive(Debug, Clone, PartialEq)]
pub struct ObliqueMercatorProjection<const C: i64> {
    proj: Rc<RefCell<Proj>>,
    store: RefCell<OmercData>,
}
impl<const C: i64> ProjectCoordinates for ObliqueMercatorProjection<C> {
    fn code(&self) -> i64 {
        C
    }
    fn name(&self) -> &'static str {
        "Oblique Mercator"
    }
    fn names() -> &'static [&'static str] {
        &[
            "Hotine_Oblique_Mercator",
            "Hotine Oblique Mercator",
            "Hotine_Oblique_Mercator_Azimuth_Natural_Origin",
            "Hotine Oblique Mercator Azimuth Natural Origin",
            "Hotine_Oblique_Mercator_Two_Point_Natural_Origin",
            "Hotine Oblique Mercator Two Point Natural Origin",
            "Hotine_Oblique_Mercator_Azimuth_Center",
            "Hotine Oblique Mercator Azimuth Center",
            "Hotine Oblique Mercator (variant A)",
            "Hotine Oblique Mercator (variant B)",
            "Oblique_Mercator",
            "Oblique Mercator",
            "omerc",
        ]
    }
}
impl<const C: i64> CoordinateStep for ObliqueMercatorProjection<C> {
    fn new(proj: Rc<RefCell<Proj>>) -> Self {
        let mut store = OmercData::default();
        {
            let proj = &mut proj.borrow_mut();

            let mut no_off = false;
            let mut lam1 = 0.;
            let mut lam2 = 0.;
            let mut phi1 = 0.;
            let mut phi2 = 0.;
            let gamma0;
            let mut con;
            let mut _f;
            let _d;
            let mut lamc = 0.;

            store.no_rot = proj.params.get(&NO_ROTATION).unwrap_or(&ProjValue::default()).bool();
            let mut alpha_c =
                proj.params.get(&AZIMUTH_PROJECTION_CENTRE).unwrap_or(&ProjValue::default()).f64();
            let alp = alpha_c != 0.;
            let mut gamma = proj
                .params
                .get(&ANGLE_RECTIFIED_TO_SKEW_GRID)
                .unwrap_or(&ProjValue::default())
                .f64();
            let gam = gamma != 0.;
            if alp || gam {
                lamc = proj
                    .params
                    .get(&LONGITUDE_OF_PROJECTION_CENTRE)
                    .unwrap_or(&ProjValue::default())
                    .f64();
                no_off = 
                    // For libproj4 compatibility
                    proj.params.get(&NO_OFF).unwrap_or(&ProjValue::default()).bool()
                    // for backward compatibility
                    || proj.params.get(&NO_UOFF).unwrap_or(&ProjValue::default()).bool();
            } else {
                lam1 = proj
                    .params
                    .get(&LONGITUDE_OF_FIRST_POINT)
                    .unwrap_or(&ProjValue::default())
                    .f64();
                phi1 = proj
                    .params
                    .get(&LATITUDE_OF_FIRST_POINT)
                    .unwrap_or(&ProjValue::default())
                    .f64();
                lam2 = proj
                    .params
                    .get(&LONGITUDE_OF_SECOND_POINT)
                    .unwrap_or(&ProjValue::default())
                    .f64();
                phi2 = proj
                    .params
                    .get(&LATITUDE_OF_SECOND_POINT)
                    .unwrap_or(&ProjValue::default())
                    .f64();
                con = fabs(phi1);

                if fabs(phi1) > FRAC_PI_2 - TOL {
                    panic!("Invalid value for lat_1: |lat_1| should be < 90°");
                }
                if fabs(phi2) > FRAC_PI_2 - TOL {
                    panic!("Invalid value for lat_2: |lat_2| should be < 90°");
                }
                if fabs(phi1 - phi2) <= TOL {
                    panic!("Invalid value for lat_1/lat_2: lat_1 should be different from lat_2");
                }
                if con <= TOL {
                    panic!("Invalid value for lat_1: lat_1 should be different from 0");
                }
                if fabs(fabs(proj.phi0) - FRAC_PI_2) <= TOL {
                    panic!("Invalid value for lat_0: |lat_0| should be < 90°");
                }
            }

            let com = sqrt(proj.one_es);
            if fabs(proj.phi0) > EPS {
                let sinph0 = sin(proj.phi0);
                let cosph0 = cos(proj.phi0);
                con = 1. - proj.es * sinph0 * sinph0;
                store.b = cosph0 * cosph0;
                store.b = sqrt(1. + proj.es * store.b * store.b / proj.one_es);
                store.a = store.b * proj.k0 * com / con;
                _d = store.b * com / (cosph0 * sqrt(con));
                _f = _d * _d - 1.;
                if _f <= 0. {
                    _f = 0.;
                } else {
                    _f = sqrt(_f);
                    if proj.phi0 < 0. {
                        _f = -_f;
                    }
                }
                _f += _d;
                store.e = _f;
                store.e *= pow(tsfn(proj.phi0, sinph0, proj.e), store.b);
            } else {
                store.b = 1. / com;
                store.a = proj.k0;
                _f = 1.;
                _d = _f;
                store.e = _d;
            }
            if alp || gam {
                if alp {
                    gamma0 = aasin(sin(alpha_c) / _d);
                    if !gam {
                        gamma = alpha_c;
                    }
                } else {
                    gamma0 = gamma;
                    alpha_c = aasin(_d * sin(gamma0));
                    if gamma <= 90. - proj.phi0 {
                        // For a sphere, |gamma| must be <= 90 - |lat_0|
                        // On an ellipsoid, this is very slightly above
                        panic!("Invalid value for gamma: given lat_0 value, |gamma| should be <= ");
                    }
                }

                if fabs(fabs(proj.phi0) - FRAC_PI_2) <= TOL {
                    panic!("Invalid value for lat_0: |lat_0| should be < 90°");
                }

                proj.lam0 = lamc - aasin(0.5 * (_f - 1. / _f) * tan(gamma0)) / store.b;
            } else {
                let _h = pow(tsfn(phi1, sin(phi1), proj.e), store.b);
                let l = pow(tsfn(phi2, sin(phi2), proj.e), store.b);
                _f = store.e / _h;
                let p = (l - _h) / (l + _h);
                if p == 0. {
                    // Not quite, but es is very close to 1...
                    panic!("Invalid value for eccentricity");
                }
                let mut j = store.e * store.e;
                j = (j - l * _h) / (j + l * _h);
                con = lam1 - lam2;
                if con < (-PI) {
                    lam2 -= TAU;
                } else if con > PI {
                    lam2 += TAU;
                }
                proj.lam0 = adjlon(
                    0.5 * (lam1 + lam2)
                        - atan(j * tan(0.5 * store.b * (lam1 - lam2)) / p) / store.b,
                );
                let denom = _f - 1. / _f;
                if denom == 0. {
                    panic!("Invalid value for eccentricity");
                }
                gamma0 = atan(2. * sin(store.b * adjlon(lam1 - proj.lam0)) / denom);
                alpha_c = aasin(_d * sin(gamma0));
                gamma = alpha_c;
            }
            store.singam = sin(gamma0);
            store.cosgam = cos(gamma0);
            store.sinrot = sin(gamma);
            store.cosrot = cos(gamma);
            store.r_b = 1. / store.b;
            store.ar_b = store.a * store.r_b;
            store.br_a = 1. / (store.ar_b);
            store.ab = store.a * store.b;
            if no_off {
                store.u_0 = 0.;
            } else {
                store.u_0 = fabs(store.ar_b * atan(sqrt(_d * _d - 1.) / cos(alpha_c)));
                if proj.phi0 < 0. {
                    store.u_0 = -store.u_0;
                }
            }
            _f = 0.5 * gamma0;
            store.v_pole_n = store.ar_b * log(tan(FRAC_PI_4 - _f));
            store.v_pole_s = store.ar_b * log(tan(FRAC_PI_4 + _f));
        }

        ObliqueMercatorProjection { proj, store: store.into() }
    }
    fn forward<P: TransformCoordinates>(&self, p: &mut P) {
        omerc_e_forward(&self.store.borrow(), &self.proj.borrow(), p);
    }
    fn inverse<P: TransformCoordinates>(&self, p: &mut P) {
        omerc_e_inverse(&self.store.borrow(), &self.proj.borrow(), p);
    }
}

/// Oblique Mercator Ellipsoidal forward project
pub fn omerc_e_forward<P: TransformCoordinates>(omerc: &OmercData, proj: &Proj, p: &mut P) {
    let mut u;
    let v;

    if fabs(fabs(p.phi()) - FRAC_PI_2) > EPS {
        let w = omerc.e / pow(tsfn(p.phi(), sin(p.phi()), proj.e), omerc.b);
        let one_div_w = 1. / w;
        let s = 0.5 * (w - one_div_w);
        let t = 0.5 * (w + one_div_w);
        let _v = sin(omerc.b * p.lam());
        let _u = (s * omerc.singam - _v * omerc.cosgam) / t;
        if fabs(fabs(_u) - 1.0) < EPS {
            panic!("Coordinate outside projection domain");
        }
        v = 0.5 * omerc.ar_b * log((1. - _u) / (1. + _u));
        let temp = cos(omerc.b * p.lam());
        if fabs(temp) < TOL {
            u = omerc.a * p.lam();
        } else {
            u = omerc.ar_b * atan2(s * omerc.cosgam + _v * omerc.singam, temp);
        }
    } else {
        v = if p.phi() > 0. { omerc.v_pole_n } else { omerc.v_pole_s };
        u = omerc.ar_b * p.phi();
    }
    if omerc.no_rot {
        p.set_x(u);
        p.set_y(v);
    } else {
        u -= omerc.u_0;
        p.set_x(v * omerc.cosrot + u * omerc.sinrot);
        p.set_y(u * omerc.cosrot - v * omerc.sinrot);
    }
}

/// Oblique Mercator Ellipsoidal inverse project
pub fn omerc_e_inverse<P: TransformCoordinates>(omerc: &OmercData, proj: &Proj, p: &mut P) {
    let u;
    let v;
    if omerc.no_rot {
        v = p.y();
        u = p.x();
    } else {
        v = p.x() * omerc.cosrot - p.y() * omerc.sinrot;
        u = p.y() * omerc.cosrot + p.x() * omerc.sinrot + omerc.u_0;
    }
    let q_p = exp(-omerc.br_a * v);
    if q_p == 0. {
        panic!("Coordinate outside projection domain");
    }
    let s_p = 0.5 * (q_p - 1. / q_p);
    let t_p = 0.5 * (q_p + 1. / q_p);
    let v_p = sin(omerc.br_a * u);
    let u_p = (v_p * omerc.cosgam + s_p * omerc.singam) / t_p;
    if fabs(fabs(u_p) - 1.) < EPS {
        p.set_lam(0.);
        p.set_phi(if u_p < 0. { -FRAC_PI_2 } else { FRAC_PI_2 });
    } else {
        p.set_phi(omerc.e / sqrt((1. + u_p) / (1. - u_p)));
        p.set_phi(phi2(pow(p.phi(), 1. / omerc.b), proj.e));
        if p.phi() == f64::MAX {
            panic!("Coordinate outside projection domain");
        }
        p.set_lam(-omerc.r_b * atan2(s_p * omerc.cosgam - v_p * omerc.singam, cos(omerc.br_a * u)));
    }
}
